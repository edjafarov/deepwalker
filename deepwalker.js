"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports.deepwalker = void 0;
function deepwalker(object) {
    var instance = {
        _value: object,
        _result: null,
        get: function (queryPath) {
            var instanceResult = __assign(__assign({ _result: null }, instance), { toValue: function (path) {
                    var _a, _b;
                    return path ? (_a = instanceResult === null || instanceResult === void 0 ? void 0 : instanceResult._result[0]) === null || _a === void 0 ? void 0 : _a.value[path] : (_b = instanceResult === null || instanceResult === void 0 ? void 0 : instanceResult._result[0]) === null || _b === void 0 ? void 0 : _b.value;
                }, filter: function (filterFn) {
                    return instanceResult.setResult(instanceResult._result.filter(filterFn));
                }, sort: function (sortFn) {
                    instanceResult._result.sort(sortFn);
                    return instanceResult;
                }, pick: function (array) {
                    instanceResult._result = instanceResult._result.map(function (el) {
                        return __assign(__assign({}, el), { value: Object.fromEntries(Object.entries(el.value)
                                .filter(function (_a) {
                                var key = _a[0];
                                return array.includes(key);
                            })) });
                    });
                    return instanceResult;
                }, flatten: function (array) {
                    if (array === void 0) { array = []; }
                    instanceResult._result = instanceResult._result.reduce(function (result, el) {
                        var arrayOfValues = Object.keys(el.value).filter(function (key) { return array.includes(key); }).map(function (key) {
                            return {
                                dimensions: el.dimensions.concat(key), value: __assign(__assign({}, Object.fromEntries(Object.entries(el.value)
                                    .filter(function (_a) {
                                    var key = _a[0];
                                    return !array.includes(key);
                                }))), { value: el.value[key] })
                            };
                        });
                        return result.concat(arrayOfValues);
                    }, []);
                    return instanceResult;
                }, concat: function (instance) {
                    instanceResult._result = instanceResult._result.concat(instance._result);
                    return instanceResult;
                }, slice: function (number) {
                    instanceResult._result = instanceResult._result.slice(0, number);
                    return instanceResult;
                }, toValues: function () {
                    return instanceResult._result.map(function (r) { return r.value; });
                }, toString: function (transformer) {
                    if (!instanceResult._result || instanceResult._result.length == 0)
                        return '';
                    return transformer(createResultsObject(instanceResult._result));
                }, convertEach: function (converter) {
                    if (!instanceResult._result || instanceResult._result.length == 0)
                        return [];
                    return instanceResult._result.map(converter);
                }, haveResults: function () {
                    if (!instanceResult._result || instanceResult._result.length == 0)
                        return false;
                    return true;
                }, toMap: function (path) {
                    return instanceResult._result.reduce(function (res, r, i) {
                        var deepResult = res;
                        r.dimensions.forEach(function (dim, i) {
                            if (!deepResult[dim] && i !== (r.dimensions.length - 1)) {
                                deepResult[dim] = {};
                            }
                            if (i === (r.dimensions.length - 1))
                                deepResult[dim] = path ? r.value[path] : r.value;
                            deepResult = deepResult[dim];
                        });
                        return res;
                    }, {});
                }, setResult: function (result) {
                    instanceResult._result = result;
                    return instanceResult;
                } });
            return instanceResult.setResult(walker(queryPath.split("."), object));
        }
    };
    return instance;
}
exports.deepwalker = deepwalker;
function createResultsObject(results) {
    var newResultsObject = __spreadArray([], results, true);
    newResultsObject.getOne = function (dims) {
        return (results.find(function (result) { return !result.dimensions.filter(function (d, i) { return d !== dims[i]; }).length; }) || {}).value;
    };
    return newResultsObject;
}
function walker(pathArray, object, result, i) {
    if (i === void 0) { i = 0; }
    if (pathArray.length == 0) {
        result[i].value = object;
        return result;
    }
    if (!result)
        result = [{ dimensions: [], value: undefined }];
    var step = pathArray.shift();
    if (step === "*") {
        if (typeof object !== 'object')
            return result;
        var baseDimensions_1 = __spreadArray([], result[i].dimensions, true);
        result.splice(i, 1);
        Object.keys(object).forEach(function (key, j) {
            var newResults = { dimensions: __spreadArray([], baseDimensions_1, true), value: undefined };
            newResults.dimensions.push(key);
            var resultIndex = result.length;
            result[resultIndex] = newResults;
            return walker(__spreadArray([], pathArray, true), object[key], result, resultIndex);
        });
        return result.filter(filterUndefineds);
    }
    return walker(pathArray, object[step], result, i).filter(filterUndefineds);
    ;
}
function filterUndefineds(res) {
    return res.value !== undefined;
}
