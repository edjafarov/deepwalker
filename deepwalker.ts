
type Deepwalker = {
  get: Function
}
type DeepwalkerResult = Deepwalker & {
  toValues: Function
}

export function deepwalker(object: any): Deepwalker {
  const instance = {
    _value: object,
    get: function (queryPath: string): DeepwalkerResult {
      return instanceResult.setResult(walker(queryPath.split("."), object));
    }

  };
  const instanceResult = {
    _result: null,
    ...instance,
    toValue: function () {
      return instanceResult?._result[0]?.value;
    },
    filter: function (filterFn) {
      return instanceResult.setResult(instanceResult._result.call.filter(instanceResult, filterFn))
    },
    sort: function (sortFn) {
      return instanceResult.setResult(instanceResult._result.call.filter(instanceResult, sortFn))
    },
    toValues: function () {
      return instanceResult._result.map(r => r.value);
    },
    toMap: function () {
      return instanceResult._result.reduce((res, r, i) => {
        let deepResult = res;
        r.dimensions.forEach((dim, i) => {
          if (!deepResult[dim] && i !== (r.dimensions.length - 1)) {
            deepResult[dim] = {};
          }
          if (i === (r.dimensions.length - 1)) deepResult[dim] = r.value;
          deepResult = deepResult[dim];
        })
        return res;
      }, {});
    },
    setResult: function (result) {
      instanceResult._result = result;
      return instanceResult;
    }

  }
  return instance;
}

type DeepResult = {
  dimensions: Array<string>,
  value: any
}

function walker(pathArray: Array<string>, object: any, result?: Array<DeepResult>, i: number = 0): Array<DeepResult> {
  if (pathArray.length == 0) {
    result[i].value = object;
    return result;
  }
  if (!result) result = [{ dimensions: [], value: undefined }];
  const step = pathArray.shift();
  if (step === "*") {
    if (typeof object !== 'object') return result;
    const baseDimensions = [...result[i].dimensions];
    result.splice(i, 1);
    Object.keys(object).forEach((key: string, j) => {
      const newResults = { dimensions: [...baseDimensions], value: undefined }
      newResults.dimensions.push(key);
      const resultIndex = result.length;
      result[resultIndex] = newResults;
      return walker([...pathArray], object[key], result, resultIndex);
    })
    return result.filter(filterUndefineds);
  }

  return walker(pathArray, object[step], result, i).filter(filterUndefineds);;
}

function filterUndefineds(res: DeepResult) {
  return res.value !== undefined;
}
