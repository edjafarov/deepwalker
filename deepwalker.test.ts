import { deepwalker } from './deepwalker'
const obj = {
  compare: {
    mobile: {
      facebook: {
        difference: 1
      },
      twitter: {
        difference: 4
      },
      linkedin: {
        difference: 7
      },
      instagram: {
        difference: 13
      },
      youtube: {
        difference: 40
      }
    },
    desktop: {
      facebook: {
        difference: 10
      },
      twitter: {
        difference: 22
      },
      linkedin: {
        difference: 7
      },
      instagram: {
        difference: 3
      },
      youtube: {
        difference: 1
      }
    }
  }
};
describe('deepwalker', () => {
  describe('should get data from object ', () => {
    test('by simple path', () => {
      expect(
        deepwalker(obj).get('compare.mobile.facebook.difference').toValue()
      ).toMatchSnapshot()
    })
    test('by mask', () => {
      expect(
        deepwalker(obj).get('compare.mobile.*.difference').toValues()
      ).toMatchSnapshot()
    })
  })
})
describe('deepwalker', () => {
  describe('should get data from object ', () => {
    test('by mask to hash', () => {
      expect(
        deepwalker(obj).get('compare.mobile.*.difference').toMap()
      ).toMatchSnapshot()
    })

    test('by mask to nested hash', () => {
      expect(
        deepwalker(obj).get('compare.*.*.difference').toMap()
      ).toMatchSnapshot()
    })
  })
})
/*
fetch(`
  compare {
    mobile {
      channel {
        sessions
        conversion rate
        avg basket
      }
    }
  }
`)
session/%
cr
// each('compare.mobile.*.difference', (v)=>{})
// get('compare.mobile.facebook.difference') > 10
filter('compare.*.*.session_share', essential_share)
function byValue(el){
  el.value
  el.dim1
  el.dim2
}

deepwalker(obj).get('compare.*.*.difference').sort(byValue).filter(essential_share).map(action)

*/