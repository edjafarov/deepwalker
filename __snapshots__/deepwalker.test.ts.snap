// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deepwalker should get data from object  by mask 1`] = `
Array [
  1,
  4,
  7,
  13,
  40,
]
`;

exports[`deepwalker should get data from object  by mask to hash 1`] = `
Object {
  "facebook": 1,
  "instagram": 13,
  "linkedin": 7,
  "twitter": 4,
  "youtube": 40,
}
`;

exports[`deepwalker should get data from object  by mask to nested hash 1`] = `
Object {
  "desktop": Object {
    "facebook": 10,
    "instagram": 3,
    "linkedin": 7,
    "twitter": 22,
    "youtube": 1,
  },
  "mobile": Object {
    "facebook": 1,
    "instagram": 13,
    "linkedin": 7,
    "twitter": 4,
    "youtube": 40,
  },
}
`;

exports[`deepwalker should get data from object  by simple path 1`] = `1`;
